src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -485,7 +487,19 @@ event::ptr primitive_inst::realloc_if_needed() {
         }
     }
 
-    bool can_reuse_buffer = _outputs[0] && updated_layout.count() <= max_output_layout_size;
+    // If the node's output memory is used as next input memory for the same node,
+    // the buffer should be allocated newly every time to prevent data corruption
+    bool output_used_as_its_input_next_iter = false;
+    if (get_node().is_type<concatenation>()) {
+        if (!can_be_optimized() && !allocation_done_by_other && get_users().front()->is_type<assign>()) {
+            output_used_as_its_input_next_iter = true;
+            updated_params._can_share_buffer = false;
```

Comment:
1. We need to check 
- User is assign && the readvalue input to this node uses same variable with the user assign. So the new pattern like readvalue => gather => assign will not need this forcing realloc. However if the gather is to be optimized, then the enforceement is needed. All in all, the essential factor to decide memory alloc enforcement is whether the input and output are using same varaiable memory. 
- BTW why do we need both output_used_As_its_input_next iter and update_params._can_sahre_buffer? Seems that only with  the params.can_sahre_buffer is enough. 
- Also it seems that we even don't need the impl_param.can_share_buffer, we can just pass need_reset flag as true. 
- Also the "need_reset" can be implemented inside the concat_inst

Reply:
As discussed, simply implemented concatenation's need_reset_output_memory() firsit and clean up unnecessary code. After fixing failing test cases, I will supplement the check logic for need_reset_output_memory()